File: week09/23/prob5.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def inorder_traversal(self, root):
        # Initialize an empty list to store the values
        result = []
        
        # Helper function to perform recursive inorder traversal
        def helper(node):
            if node:
                helper(node.left)  # Traverse left subtree
                result.append(node.val)  # Store the current node's value
                helper(node.right)  # Traverse right subtree
        
        # Call the helper function with the root node
        helper(root)
        return result

    def kthSmallest(self, root, k):
        """
        :type root: Optional[TreeNode]
        :type k: int
        :rtype: int
        """
        # Get the inorder traversal result, which will be sorted
        res = self.inorder_traversal(root)
        # Return the k-th smallest element (adjust for 0-based index)
        return res[k - 1]

File: week09/5/prob5.cpp
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> values;
        dfs_inorder(root, values);
        return values[k-1];
    }

    void dfs_inorder(TreeNode* curr, vector<int>& values) {
        // base case (nullptr) : do nothing
        if (curr == nullptr);
        // recursive case
        else {
            dfs_inorder(curr->left, values);
            values.push_back(curr->val);
            dfs_inorder(curr->right, values);
        }
    }
};